src/ixdat/readers/cinfdata_db.py

```diff
@@ -0,0 +1,411 @@
+        self.mass_scans = False
+        self.spectrum_list = []
+
+        plugins.activate_cinfdata()
+
+    def read(self, path_to_file, name=None, cls=None, units=None, **kwargs):
+        """Return a xx-Measurement or Spectrum with the data and metadata recorded from
+        a setup at SurfCat at given timestamp
+
+        All attributes of this reader can be accessed from the
+        measurement as `measurement.reader.attribute_name`.
+
+        Args:
+            path_to_file (str): Named argument from Measurement Class.
+                Can be used as the setup name in the cinfdatabase
+            **kwargs (dict): Key-word arguments are passed to cinf Measurement.__init__
+                setup_name (str): The setup name in the database default to path_to_file
+                timestamp (str): Timestamp the measurement started given as
+                    (YYYY-MM-DD HH:MM:SS)
+        """
+
+        self.measurement_class = kwargs.pop("measurement_class", cls)
+        self.setup_name = kwargs.pop("setup_name", path_to_file)
+        self.timestamp = kwargs.pop("timestamp", None)
+        self.comment = kwargs.pop("comment", None)
+        self.mass_scans = kwargs.pop("include_mass_scans", False)
+        self.verbose = kwargs.pop("verbose", False)
+        self.grouping_column = kwargs.pop("group", None)
+
+        # figure out whether to collect data group by 'comment' or 'timestamp'
+        if not self.grouping_column:
+            if self.timestamp and not self.comment:
+                self.grouping_column = "time"
+                self.token = self.timestamp
+            elif self.comment and not self.timestamp:
+                self.grouping_column = "comment"
+                self.token = self.comment
+            else:
+                warnings.warn(
+                    "Both a comment and a timestamp is given "
+                    "but no explicit grouping column is set. \n"
+                    f"Defaults to 'time' using '{self.timestamp}'",
+                    stacklevel=2,
+                )
+                self.grouping_column = "time"
+                self.token = self.timestamp
+
+
+        if issubclass(self.measurement_class, Spectrum):
+            return self.read_spectrums()
+
+        elif issubclass(self.measurement_class, Measurement):
+            obj_as_dict = self.read_ms()
+            if issubclass(cls, self.measurement_class):
+                self.measurement_class = cls
+            obj_as_dict.update(kwargs)
+            self.measurement = self.measurement_class.from_dict(obj_as_dict)
+
+            if self.mass_scans:
+                if self.verbose:
+                    print("adding mass scans to the measurement")
+                self.measurement = self.add_mass_scans()
+
+            # self.data_has_been_fetch = True
+
+            return self.measurement
+
+    def read_ms(self):
+        """Download MS data from cinfdata_database"""
+
+        self.cinf_db = plugins.cinfdata.connect(
```

Comment:
... or close the connection? Must admit I don't know what's best here. I have found it useful sometimes to check properties of `measurement.reader` of the returned Measurement, to see if it read well (or why not), but it does sound like there could be some unwanted side-effects in this case.

Reply:
I think that's exactly what we meant :)

Reply:
I tried to use `with`

The attribute error `__enter__` came.

I think the plugin should be rewritten to have a `__enter__` and a `__exit__` to enter and exit the connection, however, I could not figure out how to do this safely, and the Cinfdata module does not offer close connection.

Reply:
Ok, so just to make sure we are on the same page about what we talk about.

**Caching of database connections:** I regret commenting on that. I was weak for a moment in my Python ways. We should not worry about that unless it becomes a documented performance problem.

**Closing cindata objects and/or making it a context manager:** Its a nice idea. Context managers are **super nice** for making sure to close/unlock/free what was previously opened/locked/allocated and so they would be nice here. So, we could make the cinfdata plugin work as a context manager.

There are two caveats.
1) The upstream cinfdata object, doesn't actually have a `close` method to close db connection, so you will have to reach inside cinfdata object to do that
2) You will need to pass arguments to the context manager. For that, you will need a little additional magic by implementing `__call__` and store the values for the context manager locally.

A solution could look something like:

```python
class _CinfData:

    def __init__(
        self,
    ):
        self.cinfdata = None
        self._managed_cinfdata_object = None
        self._context_manager_kwargs = None

    def activate(self):
        from cinfdata import Cinfdata


        self.DB = Cinfdata

    def connect(self, setup_name=None, grouping_column=None):
        return self.DB(setup_name=setup_name, grouping_column=grouping_column)

	def __call__(self, **kwargs):
		"""... arguments here same as `connect`"""
		self._context_manager_kwargs = kwargs
        return self
 
	def __enter__(self):
		if self._managed_cinfdata_object:
			self._context_manager_kwargs = None
			raise RuntimeError(
				"Using the cinfdata plugin as a context manager more than once at the same time is not supported"
			)

		self._managed_cinfdata_object = self.connnect(**self._context_manager_kwargs)
		return self._managed_confdata_object

	def __exit__( self, exc_type, exc_val, exc_tb ):
		self._managed_cinfdata_object.connection.close()
		self._managed_cinfdata_object = None
		self._context_manager_kwargs = None
```

(A little note on the magic. You can make an object callable, by implementing `__call__`. It stores the kwargs for connect and then returns self...! because, then after being called, the `with` machinery will call `__enter__` on it. This will need commenting to pass code review :D Note also, that the arguments for the `__exit__` method is all about exceptions. Context manageres will trap exceptions from the code executed within them, to make sure that the cleanup code in `__exit__` is always done (and I think re-raise them afterwards, you will have to check that). But passing the exception data in, if any, will allow you to respond to it if you need to)

It would then be used like this:
```python
with plugins.cinfdata(...kwargs_for_connect...) as cinf_db:
    ...
```
**On single use reader:** This is a comment referring specifically to the ixdat reader. What this pertains to is the questions about whether it should be possible to call `read` more than once on an ixdat reader object. The point of the discussion is that these reader object often have **a lot** of state (all the properties), like yours. All of this state makes for a large error potential, with the sort of error where some state is left over from one call to `read` to the next, creating ixdat objects with data that is based on mixed information and is therefore completely broken. To remove this error potential in an easy way, we usually make reader single use, so that we can never have mixed/left over information. The way we commonly make them single use is that the very first thing `read` does, is to set/check a boolean for whether it has been used, and if it has, raise an exception explaining that the readers are single use and to create a new one.